python/cudf_polars/cudf_polars/experimental/single.py

```diff
@@ -0,0 +1,216 @@
+"""Single-partition LogicalPlan nodes."""
+
+from __future__ import annotations
+
+from functools import cached_property, singledispatch
+from typing import TYPE_CHECKING
+
+from cudf_polars.dsl.ir import (
+    IR,
+    Cache,
+    DataFrameScan,
+    Distinct,
+    Filter,
+    GroupBy,
+    HConcat,
+    HStack,
+    Join,
+    MapFunction,
+    Projection,
+    PythonScan,
+    Reduce,
+    Scan,
+    Select,
+    Slice,
+    Sort,
+    Union,
+)
+from cudf_polars.dsl.traversal import CachingVisitor
+from cudf_polars.experimental.parallel import PartitionInfo
+
+if TYPE_CHECKING:
+    from cudf_polars.dsl.ir import IR
+
+
+class SPartitionwise:
+    """Single partition-wise PartitionedIR."""
+
+    @cached_property
+    def _key(self):
+        return f"{type(self).__name__.lower()}-{hash(self)}"
+
+    def _tasks(self):
+        return {
+            (self._key, 0): (
+                self.do_evaluate,
+                *self._non_child_args,
+                *((child._key, 0) for child in self.children),
+            )
+        }
+
+    @cached_property
+    def _parts(self) -> PartitionInfo:
+        return PartitionInfo(npartitions=1)
+
+
+class SParPythonScan(PythonScan, SPartitionwise):
+    """Single-partition demo class."""
+
+
+class SParScan(Scan, SPartitionwise):
+    """Single-partition demo class."""
+
+
+class SParCache(Cache, SPartitionwise):
+    """Single-partition demo class."""
+
+
+class SParDataFrameScan(DataFrameScan, SPartitionwise):
+    """Single-partition demo class."""
+
+
+class SParSelect(Select, SPartitionwise):
+    """Single-partition demo class."""
+
+
+class SParReduce(Reduce, SPartitionwise):
+    """Single-partition demo class."""
+
+
+class SParGroupBy(GroupBy, SPartitionwise):
+    """Single-partition demo class."""
+
+
+class SParJoin(Join, SPartitionwise):
+    """Single-partition demo class."""
+
+
+class SParHStack(HStack, SPartitionwise):
+    """Single-partition demo class."""
+
+
+class SParDistinct(Distinct, SPartitionwise):
+    """Single-partition demo class."""
+
+
+class SParSort(Sort, SPartitionwise):
+    """Single-partition demo class."""
+
+
+class SParSlice(Slice, SPartitionwise):
+    """Single-partition demo class."""
+
+
+class SParFilter(Filter, SPartitionwise):
+    """Single-partition demo class."""
+
+
+class SParProjection(Projection, SPartitionwise):
+    """Single-partition demo class."""
+
+
+class SParMapFunction(MapFunction, SPartitionwise):
+    """Single-partition demo class."""
+
+
+class SParUnion(Union, SPartitionwise):
+    """Single-partition demo class."""
+
+
+class SParHConcat(HConcat, SPartitionwise):
+    """Single-partition demo class."""
+
+
+@singledispatch
+def _single_partition_node(node: IR, rec) -> SPartitionwise:
+    raise NotImplementedError(f"Cannot convert {type(node)} to PartitionedIR.")
+
+
+@_single_partition_node.register
+def _(node: PythonScan, rec) -> SParPythonScan:
+    return SParPythonScan(*node._ctor_arguments(map(rec, node.children)))
```

Comment:
@wence- - This PR is still rough, so it may not make sense to debug type-hinting problems yet. That said, I'm getting a ton of errors that I don't know how to address like: `error: Argument 1 to "SParPythonScan" has incompatible type "*Sequence[Any | IR]"; expected "Mapping[str, Any]"  [arg-type]`

Reply:
Ah, yeah, can you apply:

```diff
diff --git a/python/cudf_polars/cudf_polars/dsl/nodebase.py b/python/cudf_polars/cudf_polars/dsl/nodebase.py
index 228d300f46..c96414ad3c 100644
--- a/python/cudf_polars/cudf_polars/dsl/nodebase.py
+++ b/python/cudf_polars/cudf_polars/dsl/nodebase.py
@@ -40,7 +40,7 @@ class Node(Generic[T]):
     children: tuple[T, ...]
     _non_child: ClassVar[tuple[str, ...]] = ()
 
-    def _ctor_arguments(self, children: Sequence[T]) -> Sequence[Any | T]:
+    def _ctor_arguments(self, children: Sequence[T]) -> Sequence[Any]:
         return (*(getattr(self, attr) for attr in self._non_child), *children)
 
     def reconstruct(
```

_But_ you should use the `reconstruct` method:

Then you can do `X.reconstruct(map(rec, node.children))` I suspect.

Reply:
I probably need to fiddle with the types a bit, perhaps this:
```
diff --git a/python/cudf_polars/cudf_polars/dsl/nodebase.py b/python/cudf_polars/cudf_polars/dsl/nodebase.py
index 228d300f46..9fc855a764 100644
--- a/python/cudf_polars/cudf_polars/dsl/nodebase.py
+++ b/python/cudf_polars/cudf_polars/dsl/nodebase.py
@@ -8,7 +8,7 @@ from __future__ import annotations
 from typing import TYPE_CHECKING, Any, ClassVar, Generic, TypeVar
 
 if TYPE_CHECKING:
-    from collections.abc import Hashable, Sequence
+    from collections.abc import Hashable, Iterable, Sequence
 
     from typing_extensions import Self
 
@@ -40,12 +40,12 @@ class Node(Generic[T]):
     children: tuple[T, ...]
     _non_child: ClassVar[tuple[str, ...]] = ()
 
-    def _ctor_arguments(self, children: Sequence[T]) -> Sequence[Any | T]:
+    def _ctor_arguments(self, children: Iterable[T]) -> Sequence[Any]:
         return (*(getattr(self, attr) for attr in self._non_child), *children)
 
     def reconstruct(
-        self, children: Sequence[T]
-    ) -> Self:  # pragma: no cover; not yet used
+        self, children: Iterable[T]
+    ) -> Self:
         """
         Rebuild this node with new children.
 
```